src/scheduler.rs

```diff
@@ -0,0 +1,663 @@
+    }
+}
+
+impl<T, W> Drop for Inner<T, W> {
+    fn drop(&mut self) {
+        // Once we're in the destructor for `Inner<T, W>` we need to clear out the
+        // mpsc queue of nodes if there's anything left in there.
+        //
+        // Note that each node has a strong reference count associated with it
+        // which is owned by the mpsc queue. All nodes should have had their
+        // items dropped already by the `Scheduler` destructor above,
+        // so we're just pulling out nodes and dropping their refcounts.
+        unsafe {
+            loop {
+                match self.dequeue() {
+                    Dequeue::Empty => break,
+                    Dequeue::Inconsistent => abort("inconsistent in drop"),
+                    Dequeue::Data(ptr) => drop(ptr2arc(ptr)),
+                }
+            }
+        }
+    }
+}
+
+impl Wakeup for AtomicTask {
+    fn wakeup(&self) {
+        self.notify();
+    }
+}
+
+impl Wakeup for Arc<ThreadNotify> {
+    fn wakeup(&self) {
+        use executor::Notify;
+        self.notify(0);
+    }
+}
+
+impl<'a, T, W> Clone for Notify<'a, T, W> {
+    fn clone(&self) -> Self {
+        Notify(self.0)
+    }
+}
+
+impl<'a, T: fmt::Debug, W: fmt::Debug> fmt::Debug for Notify<'a, T, W> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Notiy").finish()
+    }
+}
+
+impl<'a, T, W: Wakeup> From<Notify<'a, T, W>> for NotifyHandle {
+    fn from(handle: Notify<'a, T, W>) -> NotifyHandle {
+        unsafe {
+            let ptr = handle.0.clone();
+            let ptr = mem::transmute::<Arc<Node<T, W>>, *mut ArcNode<T, W>>(ptr);
+            NotifyHandle::new(hide_lt(ptr))
+        }
+    }
+}
+
+struct ArcNode<T, W>(PhantomData<(T, W)>);
+
+// We should never touch `T` on any thread other than the one owning
+// `Scheduler`, so this should be a safe operation.
+//
+// `W` already requires `Sync + Send`
+unsafe impl<T, W: Wakeup> Send for ArcNode<T, W> {}
+unsafe impl<T, W: Wakeup> Sync for ArcNode<T, W> {}
+
+impl<T, W: Wakeup> executor::Notify for ArcNode<T, W> {
+    fn notify(&self, _id: usize) {
+        unsafe {
+            let me: *const ArcNode<T, W> = self;
+            let me: *const *const ArcNode<T, W> = &me;
+            let me = me as *const Arc<Node<T, W>>;
+            Node::notify(&*me)
+        }
+    }
+}
+
+unsafe impl<T, W: Wakeup> UnsafeNotify for ArcNode<T, W> {
+    unsafe fn clone_raw(&self) -> NotifyHandle {
+        let me: *const ArcNode<T, W> = self;
+        let me: *const *const ArcNode<T, W> = &me;
+        let me = &*(me as *const Arc<Node<T, W>>);
+        Notify(me).into()
+    }
+
+    unsafe fn drop_raw(&self) {
+        let mut me: *const ArcNode<T, W> = self;
+        let me = &mut me as *mut *const ArcNode<T, W> as *mut Arc<Node<T, W>>;
+        ptr::drop_in_place(me);
+    }
+}
+
+unsafe fn hide_lt<T, W: Wakeup>(p: *mut ArcNode<T, W>) -> *mut UnsafeNotify {
+    mem::transmute(p as *mut UnsafeNotify)
```

Comment:
I think this `transmute` call isn't actually doing anything?

Reply:
It is hiding the lifetime :)

That said, I think that `W` needs to be `'static`.